### PR TITLE
Added permissions by roles

### LINKED ISSUE #173: Permissions
We (currently) have login via username and password.

If it only applies to a limited number of users, this isn't a burden.

I do believe it is possible to have a workable system where most users don't need to login - removing the burden of integrating some external authorisation system, or having another username/password for all staff.

_Eventually_ the user roles will come in with the HTTP properties.  As a substitute for that, we will add roles to our `users` table.

We are going to have a `role-based` permission model.  We will have these two roles:
- `rco-user`
- `rco-power-user`

We will modify the `users` table to have a `roles` field. This will contain an array of strings.  The `UserList` form will show these strings in a series of lozenges/badges.  The `UserEdit` form will use a multi-select drop-down to allow these string items to be edited.

Note: when we make this change, we will remove the `admin` property from `User`.

## Users (no authentication)
- view lists of `projects`, `batches`, `items` (including item loan state)
- ability to filter data by all possible fields
- ability to control visible columns in all list views
- ability to (eventually) submit a request from the `Items` page relating to a selected set of items
- view list of users
- create a muster list for a user

## `rco-user` (username/password)
- as above, plus:
- add/edit `projects`, `batches`, `items`, `platforms`
- loan/return items
- create destruction certificate for items
- despatch items
- change item locations
- create muster forms

## `rco-power-user` (username/password)
- as above, plus:
- delete `projects`, `batches`, `items`
- add/edit/delete other `ReferenceData` tables, notably including:
    - users
    - protective marking lists

### PR DESCRIPTION
**Pull request by salony75way**
fixes #173

### CODE REVIEW

File: src/providers/authProvider/permissions.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+type UserRoles = UserRole | 'user'
```
Comment (IanMayo) [inline src/providers/authProvider/permissions.ts:1]: This is an impressive module @salony75way - you've clearly done this before 🤣 

I wouldn't have had a clue where to start ;-)

**Comment by IanMayo** [comment]
I've had a play @salony75way . Two changes please:
1.  to start off with, all users should be able to read all reference data.
2. we need to allow `unregistered` users to browse the data without logging in. So, they won't have user accounts.

File: src/providers/authProvider/permissions.ts (hunk 1 above)
Comment (salony75way) [reply src/providers/authProvider/permissions.ts:1]: @IanMayo  Yes i did this once, But in react admin we have Role-Based Access Control (RBAC) library but its only in  enterprise edition so we have to create it custom.

**Comment by IanMayo** [comment]
Thanks @salony75way .  I can't "create" items as either `rco-user` or `power-user`.

I _think_ we should  introduce `create` permission, in addition to `write`.

**Comment by IanMayo** [comment]
Hello @salony75way - while it appeared broken, after a rebuild, followed by logout/login, it seems to be working.

But, IMHO we need some brief unit tests for these methods:
- `protectedRoutes`
- `useCanAccess`

That will make it easier to test edge cases, and spot anything that may break in the future.

**Comment by salony75way** [comment]
> 

As of now we are including `update` permission in `write`, So please let me know if we need only `create` permission here or i need to introduce a new permission (`update`)?

**Comment by IanMayo** [comment]
> As of now we are including `update` permission in `write`, So please let me know if we need only `create` permission here or i need to introduce a new permission (`update`)?

Hmm, I've been having a play.  The current levels of permission do seem ok. So no changes to levels of permission :-)

File: src/providers/authProvider/permissions.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    platforms: { read: true, write: true, delete: false },
+    'reference-data': { read: true, write: true, delete: false }
+  },
+  'rco-power-user': {
+    '*': { all: '*' }
+  }
+}
+
+export default permissions
+
```
Comment (IanMayo) [inline src/providers/authProvider/permissions.ts:24]: @salony75way - for the following method, I wish to understand the two parameters.

Is the first parameter the set of permissions being requested, and the second parameter the responses to those requests?

**Comment by IanMayo** [comment]
One small change please. Could we only who the `Edit` button for a Show form if the user has write permission?  Currently it jumps to the relevant `List` view, but IMHO it would be better if the Edit button was either deleted or disabled if the user doesn't have `write` permission for that resource.

File: src/providers/authProvider/permissions.ts (hunk 2 above)
Comment (salony75way) [reply src/providers/authProvider/permissions.ts:24]: Hi @IanMayo,
`generatePermission` method is a helper for `mergedPermission` to merge permission if user has multiple permissions.

`generatePermission` method takes two arguments, first is permission for resource  and second is merged permissions object.

**Review by IanMayo — APPROVED** [review]
